### PR TITLE
Update default TrimMode for WebSDK projects

### PR DESCRIPTION
**Pull request by captainsafia**
Set `TrimMode=full` for web projects where `PublishTrimmed=true`.

Note: we could achieve the same effect by removing the property all together and relying on the defaults the linker targets set but I like the explicitness of this approach.

### CODE REVIEW

**Comment by DamianEdwards** [comment]
@captainsafia might be worth getting @agocke to check this too. Also, just to confirm, this should have on impact on Blazor WebAssembly projects right? They have to still use `<TrimMode>partial</TrimMode>` as I understand it.

**Comment by captainsafia** [comment]
>  might be worth getting @agocke to check this too. 

Sure.

> Also, just to confirm, this should have on impact on Blazor WebAssembly projects right? They have to still use <TrimMode>partial</TrimMode> as I understand it.

Yes, Blazor is unaffected. The fact that [these tests pass](https://github.com/dotnet/sdk/blob/da36c9856a12b1f63ed1053520426a8632db30d8/src/Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests/WasmPublishIntegrationTest.cs) helps us confirm this.

**Comment by eerhardt** [comment]
> Also, just to confirm, this should have on impact on Blazor WebAssembly projects right? They have to still use `<TrimMode>partial</TrimMode>` as I understand it.

Blazor WASM sets it early in the project (in the .props file) here:

https://github.com/dotnet/sdk/blob/6446f88894017e0143c6bc1a5c017ee882333e84/src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.props#L20-L23

File: src/WebSdk/Publish/Targets/Microsoft.NET.Sdk.Publish.targets [hunk 1]
```diff
@@ -94,9 +94,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
 
     <EFSQLScriptsFolderName Condition="$(EFSQLScriptsFolderName) == ''">EFSQLScripts</EFSQLScriptsFolderName>
 
-    <!-- If TrimMode has not already been set, default to partial trimming, as AspNet is not currently fully trim-compatible -->
-    <!-- We allow the full trimming default with PublishAot enabled, because the parts that are not trim-compatible are not aot-compatible either -->
-    <TrimMode Condition="'$(TrimMode)' == '' and '$(PublishTrimmed)' == 'true' and '$(PublishAot)' != 'true'">partial</TrimMode>
+    <TrimMode Condition="'$(TrimMode)' == '' and '$(PublishTrimmed)' == 'true'">full</TrimMode>
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/WebSdk/Publish/Targets/Microsoft.NET.Sdk.Publish.targets:97]: From [here](https://github.com/dotnet/sdk/pull/33962#issue-1801520029):

> Note: we could achieve the same effect by removing the property all together and relying on the defaults the linker targets set but I like the explicitness of this approach.

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/WebSdk/Publish/Targets/Microsoft.NET.Sdk.Publish.targets:97]: I don't agree with this reasoning. The underlying intention here is to use the default. If the trimming team decides to make another default in the future, we should use that then.

We don't explicitly set the defaults of all the other MSBuild properties - `Configuration`, `DebugType`, etc. Why would TrimMode be any different?

[the remark this answers is not in the document]
Comment (captainsafia) [reply src/WebSdk/Publish/Targets/Microsoft.NET.Sdk.Publish.targets:97]: > The underlying intention here is to use the default. If the trimming team decides to make another default in the future, we should use that then.

Relying on the default in that ways means that we won't have a gate for validating that things don't break should a new TrimMode be adopted. Is that something we're OK with?

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/WebSdk/Publish/Targets/Microsoft.NET.Sdk.Publish.targets:97]: I'm OK with it. That would be the reasonability of the people that are changing the default.

Another reason to not mess with explicitly setting the default is if some other MSBuild logic wants to change the default, they only need to consider where the default happens in the ILLink targets. But if we explicitly set the default here as well - that other MSBuild logic would need to figure out how to handle Web apps as well.

[the remark this answers is not in the document]
Comment (captainsafia) [reply src/WebSdk/Publish/Targets/Microsoft.NET.Sdk.Publish.targets:97]: > Another reason to not mess with explicitly setting the default is if some other MSBuild logic wants to change the default, they only need to consider where the default happens in the ILLink targets. But if we explicitly set the default here as well - that other MSBuild logic would need to figure out how to handle Web apps as well.

This is a good point.

**Review by eerhardt — APPROVED** [review]

**Review by agocke — APPROVED** [review]
LGTM, thanks!

Blazor may be worth a test somewhere -- I assume there's one in their repo.

**Comment by captainsafia** [comment]
> LGTM, thanks!
> 
> Blazor may be worth a test somewhere -- I assume there's one in their repo.

Yes, [here](https://github.com/dotnet/sdk/blob/da36c9856a12b1f63ed1053520426a8632db30d8/src/Tests/Microsoft.NET.Sdk.BlazorWebAssembly.Tests/WasmPublishIntegrationTest.cs).

File: src/WebSdk/Publish/Targets/Microsoft.NET.Sdk.Publish.targets [hunk 2]
```diff
@@ -93,10 +93,6 @@ Copyright (C) Microsoft Corporation. All rights reserved.
     <PublishIntermediateOutputPath Condition="'$(DockerPublish)' == 'true'">obj/Docker/publish/</PublishIntermediateOutputPath>
 
     <EFSQLScriptsFolderName Condition="$(EFSQLScriptsFolderName) == ''">EFSQLScripts</EFSQLScriptsFolderName>
-
-    <!-- If TrimMode has not already been set, default to partial trimming, as AspNet is not currently fully trim-compatible -->
-    <!-- We allow the full trimming default with PublishAot enabled, because the parts that are not trim-compatible are not aot-compatible either -->
-    <TrimMode Condition="'$(TrimMode)' == '' and '$(PublishTrimmed)' == 'true' and '$(PublishAot)' != 'true'">partial</TrimMode>
```
Comment (eerhardt) [inline src/WebSdk/Publish/Targets/Microsoft.NET.Sdk.Publish.targets:99]: Uggh, I just realized that we probably should be leaving `TrimMode=partial` for TFMs below `net8.0`. I don't think we should be changing behavior for people when they upgrade their SDK. Only when they retarget from `net6/7.0` => `net8.0`.

**Pull request merged**